src/gConsent/query/query.ts

```diff
@@ -0,0 +1,262 @@
+  | "Expired"
+  | "Active"
+  | "Revoked";
+
+/**
+ * Supported Access Credential types
+ */
+export type CredentialType =
+  | "SolidAccessRequest"
+  | "SolidAccessGrant"
+  | "SolidAccessDenial";
+
+/**
+ * Supported durations for Access Credential filtering.
+ */
+export const DURATION = {
+  ONE_DAY: "P1D",
+  ONE_WEEK: "P7D",
+  ONE_MONTH: "P1M",
+  THREE_MONTHS: "P3M",
+} as const;
+
+export type CredentialFilter = {
+  /**
+   * The Access Credential type (e.g. Access Request, Access Grant...)
+   */
+  type?: CredentialType;
+  /**
+   * The Access Credential status (e.g. Active, Revoked...)
+   */
+  status?: CredentialStatus;
+  /**
+   * WebID of the Agent who issued the Access Credential.
+   */
+  fromAgent?: UrlString;
+  /**
+   * WebID of the Agent who is the Access Credential recipient.
+   */
+  toAgent?: UrlString;
+  /**
+   * URL of the resource featured in the Access Credential.
+   */
+  resource?: UrlString;
+  /**
+   * URL of the Access Credential purpose.
+   */
+  purpose?: UrlString;
+  /**
+   * Interval (expressed using ISO 8601) during which the Credential was issued.
+   */
+  issuedWithin?: "P1D" | "P7D" | "P1M" | "P3M";
+  /**
+   * Interval (expressed using ISO 8601) during which the Credential was revoked.
+   */
+  revokedWithin?: "P1D" | "P7D" | "P1M" | "P3M";
+  /**
+   * Number of items per page.
+   */
+  pageSize?: number;
+  /**
+   * Target page (for result pagination).
+   */
+  page?: string;
+};
+
+const FILTER_ELEMENTS: Array<keyof CredentialFilter> = [
+  "fromAgent",
+  "issuedWithin",
+  "page",
+  "pageSize",
+  "purpose",
+  "resource",
+  "revokedWithin",
+  "status",
+  "toAgent",
+  "type",
+] as const;
+
+// The following ensures that, when we add a value to the FILTER_ELEMENTS array, TS complains
+// if there are missing cases when handling the value.
+type SupportedFilterElements =
+  typeof FILTER_ELEMENTS extends Array<infer E> ? E : never;
+
+// The type assertion is okay in the context of the type guard.
+const isSupportedFilterElement = (x: unknown): x is SupportedFilterElements =>
+  FILTER_ELEMENTS.includes(x as keyof CredentialFilter);
+
+const PAGING_RELS = ["first", "last", "prev", "next"] as const;
+
+export type CredentialResult = {
+  /**
+   * Page of Access Credentials matching the query.
+   */
+  items: DatasetWithId[];
+  /**
+   * First page of query results.
+   */
+  first?: CredentialFilter;
+  /**
+   * Previous page of query results.
+   */
+  prev?: CredentialFilter;
+  /**
+   * Next page of query results.
+   */
+  next?: CredentialFilter;
+  /**
+   * Last page of query results.
+   */
+  last?: CredentialFilter;
+};
+
+function toCredentialFilter(url: URL): CredentialFilter {
+  const filter: CredentialFilter = {};
+  url.searchParams.forEach((value, key) => {
+    if (isSupportedFilterElement(key)) {
+      Object.assign(filter, { [`${key}`]: value });
+    }
+  });
+  return filter;
+}
+
+function hasItems(x: unknown): x is { items: unknown[] } {
```

Comment:
This appears to test that items exists and is an array, so `hasItems` means it has the items field correctly defined as opposed to there are items present in the result. Is that confusing? I initially thought it was going to include `items.length > 0` in the result.

Reply:
Just to make sure I understand the confusion correctly, what creates confusion is that the name of the function suggests we are checking for a non-zero array, when it is only a type guard? I don't mind renaming it to anything else, it's just an internal function anyway. The complete signature showing that it i a type guard, I think it may limit the confusion, but I see your point.

I've seen two conventions for type guards: `hasX` when checking for a single attribute (like here) and `isX` when checking for multiple attributes making up a type. This could be renamed `isQueryResult` for instance, to remain idiomatic TS.

Reply:
You've understood my concern and it comes from someone not used to type guards, so I'm OK if you think it is not an issue. I thought you might respond as you did which suggests that with a little more experience I would have had no question about this.

Reply:
Refactoring this in light of this comment, I actually ended up removing the type guards: 87c5e87e8d16028db4f11d0dc54ae488704ad08b